truffleruby/src/main/java/org/truffleruby/language/dispatch/CachedDispatchNode.java

```diff
@@ -63,20 +63,32 @@ public CachedDispatchNode(
         this.next = next;
     }
 
+    protected boolean sendingFrames() {
+        return sendsFrame || readCaller != null;
+    }
+
     public void replaceSendingFrame() {
-        CompilerDirectives.transferToInterpreterAndInvalidate();
+        assert readCaller == null;
         if (!sendsFrame) {
-            sendsFrame = true;
+            replaceSendingFrame(true, null);
         }
     }
 
     public void replaceSendingCallerFrame(CallerFrameAccess access) {
-        CompilerDirectives.transferToInterpreterAndInvalidate();
+        assert sendsFrame == false;
         if (readCaller == null) {
-            readCaller = new ReadCallerFrameNode(access);
+            replaceSendingFrame(false, new ReadCallerFrameNode(access));
```

Comment:
No. there are two possible frames a dispatch node could send, its own (if it is a normal method) or it's caller's (if it is a `#send`. The `false` refers to its frame since we've been asked to send the caller. There are also asserts in the two methods `replaceSendingFrame` and `replaceSendingCallerFrame` as a safeguard against requesting both.